src/app/dao/tests/PartDAO.spec.ts

```diff
@@ -60,43 +74,249 @@ describe('PartDAO', () => {
+        // Given a non-verified user
+        const token: string = '{"sub": "bar@bar.com", "email": "bar@bar.com", "email_verified": false}';
+        const credential: FireAuth.UserCredential =
+            await FireAuth.signInWithCredential(TestBed.inject(FireAuth.Auth),
+                                                FireAuth.GoogleAuthProvider.credential(token));
+        await userDAO.set(credential.user.uid, { verified: false, username: 'user' });
+
+        // When creating a part
+        const result: Promise<string> = partDAO.create(PartMocks.INITIAL);
+        // Then it should fail
+        await expectFirebasePermissionDenied(result);
+    });
+    it('should forbid player to change playerZero/playerOne/beginning once a part has started', async() => {
+        // Given a part that has started (i.e., turn >= 0), and a player (here creator)
+        const creatorUser: FireAuth.User = await createConnectedGoogleUser('foo@bar.com', 'creator');
+        const creator: MinimalUser = { id: creatorUser.uid, name: 'creator' };
+        await partDAO.create({
+            ...PartMocks.INITIAL,
+            turn: 1,
+            playerZero: creator,
+            playerOne: UserMocks.OPPONENT_MINIMAL_USER,
+        });
+
+        const updates: Partial<Part>[] = [
+            { playerZero: UserMocks.OPPONENT_MINIMAL_USER },
+            { playerOne: creator },
+            { beginning: serverTimestamp() },
+        ];
+        for (const update of updates) {
+            // When trying to change the field
+            const result: Promise<void> = partDAO.update('partId', update);
+            // Then it should fail
+            await expectFirebasePermissionDenied(result);
+        }
+    });
+    it('should forbid non-player to change writable fields', async() => {
+        // Given a part, and a user who is not playing in this game
+        await createConnectedGoogleUser('foo@bar.com', 'creator');
+        await partDAO.create(PartMocks.INITIAL);
+        await signOut();
+        await createConnectedGoogleUser('bar@bar.com', 'non-player');
+
+        const updates: Partial<Part>[] = [
+            { lastUpdate: { index: 1, player: 0 } },
+            { turn: 42 },
+            { result: 3 },
+            { listMoves: [{ a: 1 }] },
+            { lastUpdateTime: serverTimestamp() },
+            { remainingMsForZero: 42 },
+            { remainingMsForOne: 42 },
+            { winner: 'me!' },
+            { loser: 'you' },
+            { scorePlayerZero: 42 },
+            { scorePlayerOne: 42 },
+            { request: Request.rematchProposed(Player.ZERO) },
+        ];
+        for (const update of updates) {
+            // When trying to change the field
+            const result: Promise<void> = partDAO.update('partId', update);
+            // Then it should fail
+            await expectFirebasePermissionDenied(result);
+        }
+    });
+    it('should forbid player to change typeGame', async() => {
```

Comment:
should forbid everybody to change typeGame ?

Reply:
ceci amène une question pour les tests: ici c'est fait pour player car on considère que player a un privilège plus haut que "anybody". Est-ce qu'on veut dupliquer le test pour player/creator/non-player/unverified-user? (et ce pour chaque test...)

Reply:
`describe('Thing to test', () => {
    it('should test what everbody should be allowed and forbid to do');
    describe('some role A rights descriptions');
    describe('some role B with no overlaps to A'', () => {
        it('should describe what A can do or cant');
        describe('some sub role of B description');
    });
});`

Un peu le pattern "mise en évidence" ?
Même si le pattern inverse est plus safe, moins de risque d'oubli ou de régression-partielle-invisible ?
J'sais pas trop

Reply:
le soucis c'est « qui est everybody » ? Si tu prends le everybody le plus général : c'est un user qui existe dans firebase mais qui est même pas vérifié. Donc à ce moment là il peut rien changer, donc il peut pas changer typeGame. Mais un player a potentiellement plus de droits qu'un unverified ou qu'un verified non player. Du coup, soit on teste pour les cas qui ont potentiellement plus de droits (player, creator), soit on teste pour tout le monde avec un peu de duplication (unverified, verified, player, creator)

Reply:
Bah, plutôt que tester tout avec tout le monde (ceci dit y a p-e un moyen de tester ça en faisant des tests multiples (chaque test loop sur chaque rôle :/ ? ) on peux faire des rôles sans overlaps: unverified, creator, chosenCandidate, unchosenCandidates, puis player/oberser sur les part en court ouais, verified lui c'pas un vrai rôle vu que on l'adapte au contexte.
Mais bon, là les tests sont p-e suffisant, on a centuplé la sécurité, soyons-po-crispy :D 8

Reply:
donc tu es satisfait là ? Ou y'a quelque chose à rajouter pour ce ticket ?

Reply:
D'abord p-e voir son avis sur mon premier point d'interrogation, que tu as ignoré !

Reply:
oui on peut dupliquer beaucoup de tests pour chaque rôle, mais ça ferait beaucoup de redondance pour pas tester grand chose de plus il me semble, je pense qu'ici on a atteint un optimum.

Reply:
là loop ferait que c'est pas beaucoup de code en plus, on peux ticketter ça pour (beaucoup?) plus tard, après tout, c'est potentiellement des choses qui sont pas testées si on affine notre code, ce serait plus futur proof ?

Reply:
ouais ça serait idéal, en faisant ça correctement on peut éviter une certaine duplication c'est sûr. Genre si on définit une hiérarchie des users (du userisme :x) on peut dire « tous les users en dessous de creator ne peuvent pas faire ceci ». On tickette ?